zebra-network/src/meta_addr.rs

```diff
@@ -79,6 +79,39 @@ impl PeerAddrState {
             AttemptPending | Responded | Failed => false,
         }
     }
+
+    /// Returns the typical connection state machine order of `self` and `other`.
+    /// Partially ordered states are sorted in connection attempt order.
+    ///
+    /// See [`MetaAddrChange::apply_to_meta_addr()`] for more details.
+    fn connection_state_order(&self, other: &Self) -> Ordering {
+        use Ordering::*;
+        match (self, other) {
+            (NeverAttemptedAlternate, NeverAttemptedAlternate)
+            | (NeverAttemptedGossiped, NeverAttemptedGossiped)
+            | (AttemptPending, AttemptPending)
+            | (Responded, Responded)
+            | (Failed, Failed) => Equal,
+            // Peers start in one of the "never attempted" states,
+            // then typically progress towards a "responded" or "failed" state.
+            //
+            // # Security
+            //
+            // Prefer gossiped addresses to alternate addresses,
+            // so that peers can't replace the addresses of other peers.
+            // (This is currently checked explicitly by the address update code,
+            // but we respect the same order here as a precaution.)
+            (NeverAttemptedAlternate, _) => Less,
+            (_, NeverAttemptedAlternate) => Greater,
+            (NeverAttemptedGossiped, _) => Less,
+            (_, NeverAttemptedGossiped) => Greater,
+            (AttemptPending, _) => Less,
+            (_, AttemptPending) => Greater,
+            (Responded, _) => Less,
+            (_, Responded) => Greater,
+            // Failed is covered by the other cases
```

Comment:
This method returns the connection state machine order, to help us choose between concurrent changes.